maggy/core/executors/dist_executor.py

```diff
@@ -0,0 +1,239 @@
+            tb_logdir, trial_log_file = _setup_logging(reporter, log_dir)
+
+            client.await_reservations()
+            config = client.get_torch_config()
+            addr, port = config["host_port"].split(":")
+            torch_config = {
+                "MASTER_ADDR": addr,
+                "MASTER_PORT": port,
+                "WORLD_SIZE": str(config["num_executors"]),
+                "RANK": str(partition_id),
+                "NCCL_BLOCKING_WAIT": "1",
+            }
+            reporter.log(f"Torch config is {torch_config}")
+
+            _setup_torch_env(torch_config)
+            _init_cluster(timeout=60, random_seed=0)
+            rank = int(torch_config["RANK"])
+            model = torch.nn.parallel.DistributedDataParallel(kwargs["model"].cuda())
+
+            reporter.log("Starting distributed training.", True)
+            # device = torch.device(torch.cuda.current_device())
+            sig = inspect.signature(train_fn)
+            if sig.parameters.get("reporter", None):
+                retval = train_fn(
+                    model=model,
+                    train_set=kwargs["train_set"],
+                    test_set=kwargs["test_set"],
+                    reporter=reporter,
+                )
+            else:
+                retval = train_fn(
+                    model=model,
+                    train_set=kwargs["train_set"],
+                    test_set=kwargs["test_set"],
+                )
+            if rank == 0:
+                retval = util._handle_return_val(
+                    retval, tb_logdir, "Metric", trial_log_file
+                )
+
+            reporter.log("Finished distributed training.", False)
+            reporter.log("Final metric: {}".format(retval), False)
+            client.finalize_metric(retval, reporter)
+        except:  # noqa: E722
+            reporter.log(traceback.format_exc(), False)
+            raise
+        finally:
+            reporter.close_logger()
+            client.stop()
+            client.close()
+
+    return wrapper_function
+
+
+def _setup_maggy_print(reporter):
+    """Modifies printing in the train function to also write to the logger.
+
+    Args:
+        reporter (Reporter): Reporter object responsible for logging.
+    """
+    builtin_print = __builtin__.print
+
+    def maggy_print(*args, **kwargs):
+        """Maggy custom print() function."""
+        builtin_print(*args, **kwargs)
+        reporter.log(" ".join(str(x) for x in args), True)
+
+    __builtin__.print = maggy_print
+
+
+def _register_with_servers(client, reporter, partition_id):
+    """Registers own address with server and starts heartbeat protocol.
+
+    Args:
+        client (Client): Client for communication with the server.
+        reporter (Reporter): Reporter responsible for heartbeat.
+        partition_id (int): Executors partition ID from Sparks RDD.
+    """
+    client_addr = client.client_addr
+    host_port = (
+        client_addr[0] + ":" + "8081"
```

Comment:
why is the port hardcoded?

Reply:
When I assigned ports outside a certain range during initial docker tests there were connection issues, so I just went with whatever worked. Gonna test if it works with all ports on the Spark cluster and change that line.